(Prepare) Update to C++20

Jojo-Schmitz opened the pull request.

igorkorsukov commented:
FYI https://github.com/musescore/MuseScore/pull/28014 
as far as I understand, they haven't solved the problem with crashpad compilation yet

Jojo-Schmitz commented:
Interesting. The code adjustments needed might but good to go though.

cbjeukendrup committed:
Fix compilation of google_crashpad_client with C++20

See https://chromium.googlesource.com/chromium/mini_chromium/+/707c87bd258dcb8c19be8a69e92efae9fd1b51ad%5E%21/

('Stolen' from #28014, commit 3)

Jojo-Schmitz commented on CMakeLists.txt:
```diff
@@ -22,7 +22,7 @@ cmake_minimum_required(VERSION 3.22) # Min. required by Qt 6.9
 project(MuseScore LANGUAGES C CXX)
 
-set(CMAKE_CXX_STANDARD 20)
+set(CMAKE_CXX_STANDARD 2a)
```
this doesn't work at all

Jojo-Schmitz commented:
Just for a test, curiosity, that all.

I found another issue (in the logs for Linux) and a hint at how to fix the errors mentioned above (in the logs for Mac)
Might fix the build issues with C++20.

Jojo-Schmitz commented:
OK, I'll add one small change more (the one I found in the Linux build) and squash them, into 3 commits

1. (Prepare) Update to C++20
2. Fix compilation of google_crashpad_client with C++20 (which isn't mine)
Let it build, so we see the remainining failures in the logs
3. Back to C++17 (dropping the intermediate test with C++2a)
Marking it ready for review

mike-spa committed:
Update to C++20 standard

'Stolen' from #28014, commit 1

mike-spa committed:
Correct QChar and QString comparisons

'Stolen' from #28014, commit 2, plus two extra fixes (for code that presumably came in later

Jojo-Schmitz committed:
More adjustments and changes needed for C++20

Jojo-Schmitz committed:
Fix gcc compiler warnings (C++20 specific)

reg.: implicit capture of ‘this’ via ‘[=]’ is deprecated in C++20 [-Wdeprecated]

Jojo-Schmitz committed:
Fix gcc compiler warnings (for third party code)

reg. (once): ignoring return value of ‘fread’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
and (twice): compound assignment with ‘volatile’-qualified left operand is deprecated [-Wvolatile]
and (once): iteration 80 invokes undefined behavior [-Waggressive-loop-optimizations]
by disabling them (third party code)

Jojo-Schmitz committed:
Revert "Fix compilation of google_crashpad_client with C++20"

This reverts commit 19b2ec42509e862d0b8b70f9ca0f1af4b3d3611f.

Jojo-Schmitz committed:
Back to C++17 standard

This reverts commit 8a221c3e20ae6cb3ab1e902ff4e4ad5a3d1e8d5a.

Jojo-Schmitz commented on buildscripts/cmake/SetupCompileWarnings.cmake:
```diff
@@ -39,6 +39,7 @@ function(target_no_warning TARGET WNAME)
         unset(CLANG_Warning)
     elseif(WNAME STREQUAL "-Wno-aggressive-loop-optimizations")
         set(MSVC_Warning )
+        set(GCC_Warning $<$<COMPILE_LANGUAGE:C>:-Wno-aggressive-loop-optimizations>)
```
doesn't help to eliminate that warning

Jojo-Schmitz commented:
rebased to resolve conflicts

Jojo-Schmitz commented:
another rebase needed :-(

igorkorsukov commented:
> another rebase needed :-(

let me do this

Jojo-Schmitz commented:
Should we close this one here then?
